api/src/main/java/jakarta/enterprise/inject/build/compatible/spi/Enhancement.java

```diff
@@ -1,36 +1,77 @@
+ * the method is called once for each constructor or method that is declared on each <em>discovered type</em>,
+ * as defined in {@link jakarta.enterprise.lang.model.declarations.ClassInfo#constructors() ClassInfo.constructors}
+ * and {@link jakarta.enterprise.lang.model.declarations.ClassInfo#methods() ClassInfo.methods}.
  * <p>
- * You can also declare a parameter of type {@link Messages Messages} to produce log messages and validation errors.
+ * If an {@code @Enhancement} method has a parameter of type {@code FieldConfig} or {@code FieldInfo},
+ * the method is called once for each field that is declared on each <em>discovered type</em>, as defined
+ * in {@link jakarta.enterprise.lang.model.declarations.ClassInfo#fields() ClassInfo.fields}.
  * <p>
- * If you need to create instances of {@link jakarta.enterprise.lang.model.types.Type Type}, you can also declare
- * a parameter of type {@link Types}. It provides factory methods for the void pseudo-type, primitive types,
- * class types, array types, parameterized types and wildcard types.
+ * Additionally, methods annotated {@code @Enhancement} may declare parameters of these types:
+ * <ul>
+ * <li>{@link Messages}</li>
+ * <li>{@link Types}</li>
+ * </ul>
  * <p>
- * If you need to create instances of {@link jakarta.enterprise.lang.model.AnnotationInfo AnnotationInfo},
- * use {@link AnnotationBuilder}.
+ * Finally, {@link AnnotationBuilder} may be used to create instances
+ * of {@link jakarta.enterprise.lang.model.AnnotationInfo AnnotationInfo}.
  *
  * @since 4.0
  */
 @Target(ElementType.METHOD)
 @Retention(RetentionPolicy.RUNTIME)
 public @interface Enhancement {
+    /**
+     * Defines the set of <em>expected types</em>. If {@link #withSubtypes() withSubtypes}
+     * is {@code true}, the set of <em>expected types</em> includes all direct and indirect
+     * subtypes of these types. If {@link #withAnnotations() withAnnotations} is defined,
+     * the set of <em>expected types</em> only includes types that use given annotations.
+     *
+     * @return the set of <em>expected types</em>
+     */
+    Class<?>[] types();
+
+    /**
+     * If {@code true}, the set of <em>expected types</em> includes all direct and
+     * indirect subtypes of given {@link #types() types}.
+     *
+     * @return whether subtypes should be included in the set of <em>expected types</em>
+     */
+    boolean withSubtypes() default false;
+
+    /**
+     * Narrows down the set of <em>expected types</em>, defined by {@link #types() types}
+     * and {@link #withSubtypes() withSubtypes}, to types that use given annotations.
+     * The annotations can appear on the type, or on any member of the type,
+     * or any parameter of any member of the type.
+     *
+     * @return types of annotations that must be present on the <em>expected types</em>
+     */
+    Class<? extends Annotation>[] withAnnotations() default Annotation.class;
```

Comment:
I'm still concerned this defaults to any annotation. It should maybe be clarified that if not specified then only types with at least one bean defining annotation will be visited?

Reply:
Also, I just noticed that Portable Extensions' `@WithAnnotations` also treats a type annotated with _given annotation_ if the type is annotated with some _other_ annotation that is, in turn, meta-annotated with _given annotation_. Do we want to add that too?

Reply:
yes stereotypes are fine